### ISSUE TITLE
False positives with option `ignore: ["consecutive-duplicates-with-different-syntaxes"]` in `declaration-block-no-duplicate-properties`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#6796]
see : https://github.com/stylelint/stylelint/issues/6632#issuecomment-1511226734

### What steps are needed to reproduce the bug?

- two subsequent declarations with the same property
- two different values but having the same AST shape and different keywords

```css
.foo {
  position: fixed;
  position: sticky;
}

.bar {
  color: color(display-p3 0 1 0);
  color: color(xyz 0 1 0);
}
```

### What Stylelint configuration is needed to reproduce the bug?

```js
 	'declaration-block-no-duplicate-properties': [
 		true,
 		{
 			ignore: [
 				'consecutive-duplicates-with-different-syntaxes',
 			],
 		},
 	],
```

### How did you run Stylelint?

js API

### Which version of Stylelint are you using?

latest

### What did you expect to happen?

I expected fewer false positives form this new option.
Some false positives are ok.

### What actually happened?

The most common false positives were two different keywords.
These are currently seen as the same syntax while they are most often fallbacks.

```css
.foo {
  width: auto;
  width: max-content; /* max-content is fairly recent feature */
}
```

### Does the bug relate to non-standard syntax?

No

### Proposal to fix the bug

I think there are multiple ways forward:

- extend the current algorithm to allow keywords with different names
- use two mappings to ignore subsequent keywords, one with common initial values (`auto`) and the css wide keywords (`inherit`), the other with common fallbacks (`fixed`)
- switch to a different algorithm that focusses more on detecting unique keywords, function names, units, delims, ... (see: https://github.com/stylelint/stylelint/issues/6632#issuecomment-1511761204) 

--------

Personally I think it would be best if we reduce the number of false positives as much as possible and make the option smarter/better over time.

The least false positives:
- distinct node types indicate different syntaxes
- distinct keywords indicate different syntaxes
- distinct function names indicate different syntaxes
- different number of child nodes indicate different syntaxes
- distinct units in dimensions indicate different syntaxes

Compared to direct comparison between values as strings this still catches a few types of errors :
- same value but one has different whitespace/comment tokens
- `10px` followed by `11px`

**Comment by ybiquitous** [comment]
@romainmenke Thanks for the report. By the way, is `consecutive-duplicates-with-different-values` in the config example you reported a mistake for `consecutive-duplicates-with-different-syntaxes`, right?

We can reproduce it with [this demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0AzCEAEwA6A7DGADhAM4CWALmRLvBimQB4xQDceBx5VNdJVYANYBPdrgC+ePEgBGAQwBOWDhkgAbCArrrNACihkShNXOEBaQgGYMABgwBGWwEoxBHVtUQNC3Y2EAvWwdnMXEQABoQBjUYADk5AFs4RBhGROMYJDASEgjwGgYAcwQQBOgAVxikVOIFChIMAF5lfAwFSpgSOhxWggByWDATBTkeXDMZDSEzXAgzKHLjMjBRmAsFCEIYOrJOvroAbRUCAgp2mHDjk56T24wyQtmFGEOru4w+yFwSGDByqgAbmsFksVhROmYAO6UAAW8zIKBQ2xguAoZhIwlRcmYJD6l16dwAuvj3uISSdiSpJBI8l8igAxTQJUYlABWuUisEIuUQNww2BA-GEMTUZFRAroApM4P4AvJAqFIrFaLpD3RFDkuCgiigEv5IGlnQoArwYXEQA).

**Comment by romainmenke** [comment]
> you reported a mistake for `consecutive-duplicates-with-different-syntaxes`, right?

right!

Was bad copy/paste, I've updated my comment :)

**Comment by ybiquitous** [comment]
@romainmenke Your proposal sounds good to me for now. Indeed, how complex will you estimate the proposed algorithm? If not too much, I guess we can accept it.

**Comment by romainmenke** [comment]
> Indeed, how complex will you estimate the proposed algorithm? If not too much, I guess we can accept it.

It is a refinement of what already exists, so should be fairly straightforward.

------

🤔 The intention of this rule and it's options is to warn users when they are writing duplicate declarations where one would never have effect.

Custom property declarations are different from regular properties because browsers never check the validity of their syntax during parsing. These are only checked when used.

Should this option also warn on this example :

```css
:root {
  --foo: block; /* will never have effect, should this warn? */
  --foo: something-new;
}

.bar {
  display: var(--foo);
}
```

**Comment by ybiquitous** [comment]
> Should this option also warn on this example :

I'm unsure why we ignore custom properties (the code seems quite old), but let's keep it as-is not to bring a new problem.

**Comment by ybiquitous** [comment]
I've labeled the issue as **ready to implement**. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by romainmenke** [comment]
> I'm unsure why we ignore custom properties (the code seems quite old), but let's keep it as-is not to bring a new problem.

I agree :)
I don't want to alter any existing logic.

I did take dashed idents into account in the update.

```css
/* previously ignored and still ignored */
:root {
  --foo: 1;
  --foo: 1; 
}
```

```css
/* same syntax and will error */
:root {
  width: var(--foo);
  width: var(--bar);
}
```
